### PR TITLE
rust: Fix build flags

### LINKED ISSUE #91191: rustup tries to update itself and fails with error: Read-only file system
**Describe the bug**

On nixpkgs master, `rustup update` tries to do a self-update and fails with

```
error: Read-only file system (os error 30) at path "/nix/store/gxx9rxzvmk4mhzamv679gc65njgc754r-rustup-1.21.1/bin/updtest9a6vzy"
```

I believe this is a regression from 20.03, where I never saw this.

**To Reproduce**
1. On nixpkgs master, `nix-env -iA rustup && rustup update`
2. Observe the above-mentioned error

**Additional context**

```
# rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: latest update on 2020-06-20, rust version 1.46.0-nightly (2d8bd9b74 2020-06-19)
info: downloading component 'rust-std' for 'wasm32-unknown-unknown'
  9.5 MiB /   9.5 MiB (100 %)   2.6 MiB/s in  3s ETA:  0s
info: downloading component 'rust-src'
info: downloading component 'cargo'
  5.0 MiB /   5.0 MiB (100 %)   2.7 MiB/s in  1s ETA:  0s
info: downloading component 'clippy'
info: downloading component 'rust-docs'
 12.2 MiB /  12.2 MiB (100 %)   2.5 MiB/s in  4s ETA:  0s
info: downloading component 'rust-std'
 16.0 MiB /  16.0 MiB (100 %)   2.7 MiB/s in  6s ETA:  0s
info: downloading component 'rustc'
 47.7 MiB /  47.7 MiB (100 %)   2.6 MiB/s in 18s ETA:  0s
info: downloading component 'rustc-dev'
 70.9 MiB /  70.9 MiB (100 %)   2.5 MiB/s in 28s ETA:  0s
info: removing previous version of component 'rust-std' for 'wasm32-unknown-unknown'
info: removing previous version of component 'rust-src'
info: removing previous version of component 'cargo'
info: removing previous version of component 'clippy'
info: removing previous version of component 'rust-docs'
info: removing previous version of component 'rust-std'
info: removing previous version of component 'rustc'
info: removing previous version of component 'rustc-dev'
info: installing component 'rust-std' for 'wasm32-unknown-unknown'
info: installing component 'rust-src'
info: installing component 'cargo'
info: installing component 'clippy'
info: installing component 'rust-docs'
 12.2 MiB /  12.2 MiB (100 %)   4.3 MiB/s in  3s ETA:  0s
info: installing component 'rust-std'
 16.0 MiB /  16.0 MiB (100 %)   9.4 MiB/s in  1s ETA:  0s
info: installing component 'rustc'
 47.7 MiB /  47.7 MiB (100 %)   9.5 MiB/s in  5s ETA:  0s
info: installing component 'rustc-dev'
 70.9 MiB /  70.9 MiB (100 %)   8.9 MiB/s in  7s ETA:  0s
error: Read-only file system (os error 30) at path "/nix/store/gxx9rxzvmk4mhzamv679gc65njgc754r-rustup-1.21.1/bin/updtest9a6vzy"

# which rustup
/run/current-system/sw/bin/rustup

# rustup --version
rustup 1.21.1 (1970-01-01)

# rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
error: Read-only file system (os error 30) at path "/nix/store/gxx9rxzvmk4mhzamv679gc65njgc754r-rustup-1.21.1/bin/updtesto9M7mI"

# rustup update --no-self-update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'

   stable-x86_64-unknown-linux-gnu unchanged - rustc 1.44.1 (c7087fe00 2020-06-17)
  nightly-x86_64-unknown-linux-gnu unchanged - rustc 1.46.0-nightly (2d8bd9b74 2020-06-19)

info: cleaning up downloads & tmp directories
```

**Notify maintainers**
@Mic92 

**Metadata**

 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.47-hardened, NixOS, 20.09.git.622cec1e30d (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`
 - nixpkgs: `/var/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### LINKED ISSUE #93119: buildRustPackage: broken installCheckPhase and ignore custom installCheckPhase
**Describe the bug**
It is a regression since #91689.

1. When `doInstallCheck = true`, `cargo test` will be run and result in infinite `Blocking waiting for file lock on build directory`.
2. The custom `installCheckPhase` is always ignored and overrided by `cargo test`.

**To Reproduce**
1. Choose any rust package using `buildRustPackage` with custom `installCheckPhase`, like `rls`.
2. Build it with `nix build -f . rls -L`.
3. The log shows that it ignores custom `installCheckPhase` and tries to run `cargo test`, but stuck in `Blocking waiting for file lock on build directory`.

**Expected behavior**
- It should print `custom check phase` and be built successfully.
- Without setting `installCheckPhase`, it should correctly run `cargo test` without infinite blocking.

**Additional context**
#91689 (may accidentally) mixes hooks, flags, phase scripts of `checkPhase` and `installCheckPhase`, and breaks some packages (including `rls` and #91359).
I think it is poor designed and should be reverted.

**Notify maintainers**

@Flakebi 
cc @zowoq  @symphorien 

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.50, NixOS, 20.09pre234194.8d05772134f (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels(root): `"nixos-20.09pre234194.8d05772134f"`
 - channels(oxa): `"home-manager, mozilla"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - buildRustPackage
  - rls
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

:warning: **Caution:** I only tested this against a few packages. Before merging this, we should ensure to not cause even further regressions.

This PR fixes a bug originally described in #91191: if a package has a special `cargoBuildFlags`-declaration, the binaries in `$out` aren't built with those since the artifacts in `target/release` get modified during `checkPhase` which doesn't use `cargoBuildFlags`.

This change copies the artifacts after `buildPhase` into a temporary location, so that the proper binaries get installed into `$out` and `cargo test` can resume the build-process in the `checkPhase`.

cc #91689
Closes #93119
Closes #91191

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by Flakebi — APPROVED** [review]
I can confirm that this fixes the rustup issue.
Some packages I tested with nixpkgs-review, all succeeded: `rustup`, `rls`, `alacritty`, `mdbook`, `ripgrep`, `rustfmt`, `sd`

File: pkgs/build-support/rust/default.nix [hunk 1]
```diff
@@ -193,13 +194,15 @@ stdenv.mkDerivation (args // {
     # This needs to be done after postBuild: packages like `cargo` do a pushd/popd in
     # the pre/postBuild-hooks that need to be taken into account before gathering
     # all binaries to install.
-    bins=$(find $releaseDir \
+    mkdir -p $tmpDir
+    cp -vr $releaseDir/* $tmpDir/
```
Comment (Ma27) [inline pkgs/build-support/rust/default.nix:198]: Before merging, the `-v` should be removed.

**Review by cole-h — APPROVED** [review]
LGTM. Tested with `tokei`, which also exhibited problems (https://github.com/NixOS/nixpkgs/pull/91298#issuecomment-647869253) -- now it shows yaml serialization in `tokei --help`!

**Comment by zowoq** [comment]
Sorry this turned into a mess and that I wasn't able to resolve it before it landed in master.

Thanks for fixing it @Ma27.

Hydra is quiet, might be worthwhile merging this to master so we aren't waiting for the next staging cycle?

**Comment by Ma27** [comment]
> Hydra is quiet, might be worthwhile merging this to master so we aren't waiting for the next staging cycle?

I don't think that it's a good idea to merge something with ~1000 rebuilds into `master`.

@vcunat would it be possible to merge this into `staging-next` to speed this up a bit as soon as this PR is approved by more folks.

**Comment by flokli** [comment]
https://github.com/NixOS/nixpkgs/pull/91090 has been approved 2 days ago - I assume `staging-next` is mostly waiting for rebuilds to finish.

**Comment by zowoq** [comment]
> #91090 has been approved 2 days ago - I assume `staging-next` is mostly waiting for rebuilds to finish.

It's already in master, the PR has been recycled a couple of times. Currently `staging-next` is empty.

**Comment by flokli** [comment]
Uh, in that case, If staging-next is empty, and this PR causes a big rebuild, shouldn't we merge it into staging, and then staging into staging-next?

**Comment by flokli** [comment]
Let's let this cook a bit in staging, and spot breakages there.

**Pull request merged**